src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -60,22 +60,26 @@ def always_uses_default_python_platform(tgt):
 
 
 def may_have_explicit_python_platform(tgt):
-  return isinstance(tgt, (PythonBinary, PythonDistribution))
+  return isinstance(tgt, PythonBinary)
 
 
 def targets_by_platform(targets, python_setup):
-  d = defaultdict(OrderedSet)
+  targets_requiring_default_platforms = []
+  explicit_platform_settings = defaultdict(OrderedSet)
   for target in targets:
     if always_uses_default_python_platform(target):
-       # There are currently no tests for this because they're super platform specific and it's hard
-       # for us to express that on CI, but https://github.com/pantsbuild/pants/issues/7616 has an
-       # excellent repro case for why this is necessary.
-      for platform in python_setup.platforms:
-        d[platform].add(target)
+      targets_requiring_default_platforms.append(target)
     elif may_have_explicit_python_platform(target):
       for platform in target.platforms if target.platforms else python_setup.platforms:
-        d[platform].add(target)
-  return d
+        explicit_platform_settings[platform].add(target)
+  # There are currently no tests for this because they're super platform specific and it's hard for
+  # us to express that on CI, but https://github.com/pantsbuild/pants/issues/7616 has an excellent
+  # repro case for why this is necessary.
+  if explicit_platform_settings:
+    for target in targets_requiring_default_platforms:
+      for platform in python_setup.platforms:
+        explicit_platform_settings[platform].add(target)
+  return dict(explicit_platform_settings)
```

Comment:
I believe this PR should work without this change, although I could make an argument that this is more correct, but I'm really not sure, and this is such a big change to the file I would *probably* be ok either completely reverting it or keeping it as long as there's a success/failure test for platform resolution (the skipped unit test, basically).